pkg/util/ssa.go

```diff
@@ -0,0 +1,46 @@
+/*
+Copyright 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"encoding/json"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// applyConfigPatch uses server-side apply to patch the object.
+type applyConfigPatch struct {
+	applyConfig interface{}
+}
+
+// Type implements Patch.
+func (p applyConfigPatch) Type() types.PatchType {
+	return types.ApplyPatchType
+}
+
+// Data implements Patch.
+func (p applyConfigPatch) Data(_ client.Object) ([]byte, error) {
+	return json.Marshal(p.applyConfig) //nolint:wrapcheck
+}
+
+// ApplyConfigPatch is a helper function for using applyConfigPatch.
+func ApplyConfigPatch(applyConfig interface{}) client.Patch {
```

Comment:
Why not use the one from controller runtime?

Reply:
```go
// applyPatch uses server-side apply to patch the object.
type applyPatch struct{}

// Type implements Patch.
func (p applyPatch) Type() types.PatchType {
	return types.ApplyPatchType
}

// Data implements Patch.
func (p applyPatch) Data(obj Object) ([]byte, error) {
	// NB(directxman12): we might technically want to be using an actual encoder
	// here (in case some more performant encoder is introduced) but this is
	// correct and sufficient for our uses (it's what the JSON serializer in
	// client-go does, more-or-less).
	return json.Marshal(obj)
}

```

```go
	if err := r.Status().Patch(ctx, mapiMachineSet, util.ApplyConfigPatch(msAc), client.ForceOwnership, client.FieldOwner("machineset-sync-controller")); err != nil {

```

IIUC we can't pass an apply config to controller runtime Apply patch implementation. 

- The Apply configs don't implement `.DeepCopy()` so don't satisfy `client.Object` (we can't use Data() then)
- We can't actually pass a patch made using `applyPatch` without specifying a raw patch and separate data. 

The if we just passed the type `client.Apply` to `Status().Patch()` the logic just passes the thing we're trying to update as `obj` to `Data()`, so we can't use ApplyConfigs - and if we don't use them we'd own all fields without `omit empty`.

Reply:
Ok, lets keep this

I wonder what the point of their implementation is then, if it clearly doesn't work 🤔

Reply:
I think it will work, but it take ownership of any fields in the resource being patched without an `omitempty` :/

Reply:
Why is our implementation any different then?